### ISSUE TITLE
Request for Detailed Parameters on LoRA Fine-Tuning Qwen2 VL 7B with Llama Factory

### ISSUE THREAD

**Issue by NJIT-Fan** [JingbiaoMei/RGCL#2]
Hi there,

I'm currently working on fine-tuning the Qwen2 VL 7B model using the llama-factory framework. I noticed your work and was wondering if you could share the specific parameters and details used in your LoRA fine-tuning comparative experiments (Harm-P).

I'm particularly interested in:

The hyperparameter settings (e.g., learning rate, batch size, number of epochs)
LoRA-specific configurations (e.g., rank, alpha, dropout rate)
Any additional configuration details that were critical to your experiments
(Optional) Details regarding the dataset and any preprocessing steps if applicable
These details would be very helpful for replicating and comparing results on my end. Thank you in advance for your time and assistance!

Best regards,

**Comment by NJIT-Fan** [comment]
**_### Consistent with prior work (Cao et al.,2022; Pramanick et al., 2021b), we merge the veryharmful and partially harmful categories into a single hateful class, while treating harmless memes as benign._**

I see, I made a mistake here, I didn't merge them

**Comment by JingbiaoMei** [comment]
Hi,

We use the default hyperparameters in the LLaMAFactory for these experiments. However, for Harm-P there is this one important thing we forgot to mention in the paper, but we will add in the camera-ready version. We use the default annotation (V1 annotation as in the original dataset) to make a fair comparison with the previous literature.  Note that if you use V2 annotation, the result will be quite different.

**Comment by NJIT-Fan** [comment]
I understand. I appreciate it.

**Comment by JingbiaoMei** [comment]
> I understand. I appreciate it.

Thanks for your interest in our work. We will shortly release the code and setup for the LMM-RGCL, stay tuned.

**Comment by NJIT-Fan** [comment]
Hi,

Sorry to bother you again.

I'm fine-tuning Qwen 2.5-VL 7B using LLaMA Factory with the default parameters(LoRA).

I've adjusted the dataset to use binary classification labels: "hateful" and "benign."

Dataset format (example):
```
{"instruction": "Task Description: Analyze the provided image along with the accompanying text to evaluate its potential harmfulness. Carefully observe all visual details and contextual clues, then decide whether the content is 'benign' (harmless) or 'hateful' (harmful).", "input": "IOWA DEMOCRATIC PARTY SAYS IT\nWILL RELEASE 'MAJORITY' OF\nCAUCUS RESULTS BY 5PM EASTERN\nPETE\nBLAMESDELAY ON CODING ISSUE' IN\nREPORTING SYSTEM\n<image>\nQuestion: Based on the image and text, is the content potentially harmful? Please answer 'benign' or 'hateful'.", "output": "hateful", "images": ["./dataset/harm/HarMeme/Harm-P/harmeme_images_us_pol/memes_1551.png"]}
```

Parameters:
```
top.booster: auto
top.checkpoint_path: []
top.finetuning_type: lora
top.model_name: Qwen2.5-VL-7B-Instruct
top.quantization_bit: none
top.quantization_method: bitsandbytes
top.rope_scaling: none
top.template: qwen2_vl
train.additional_target: ''
train.apollo_rank: 16
train.apollo_scale: 32
train.apollo_target: all
train.apollo_update_interval: 200
train.badam_mode: layer
train.badam_switch_interval: 50
train.badam_switch_mode: ascending
train.badam_update_ratio: 0.05
train.batch_size: 2
train.compute_type: bf16
train.create_new_adapter: false
train.cutoff_len: 2048
train.dataset:
- harm_p_dataset_train
train.dataset_dir: dataset/harm/HarMeme/Harm-P
train.ds_offload: false
train.ds_stage: none
train.extra_args: '{"optim": "adamw_torch"}'
train.freeze_extra_modules: ''
train.freeze_trainable_layers: 2
train.freeze_trainable_modules: all
train.galore_rank: 16
train.galore_scale: 2
train.galore_target: all
train.galore_update_interval: 200
train.gradient_accumulation_steps: 8
train.learning_rate: 5e-5
train.logging_steps: 5
train.lora_alpha: 16
train.lora_dropout: 0
train.lora_rank: 8
train.lora_target: ''
train.loraplus_lr_ratio: 0
train.lr_scheduler_type: cosine
train.mask_history: false
train.max_grad_norm: '1.0'
train.max_samples: '100000'
train.neat_packing: false
train.neftune_alpha: 0
train.num_train_epochs: '3.0'
train.packing: false
train.ppo_score_norm: false
train.ppo_whiten_rewards: false
train.pref_beta: 0.1
train.pref_ftx: 0
train.pref_loss: sigmoid
train.report_to:
- none
train.resize_vocab: false
train.reward_model: []
train.save_steps: 100
train.swanlab_api_key: ''
train.swanlab_link: ''
train.swanlab_mode: cloud
train.swanlab_project: llamafactory
train.swanlab_run_name: ''
train.swanlab_workspace: ''
train.train_on_prompt: false
train.training_stage: Supervised Fine-Tuning
train.use_apollo: false
train.use_badam: false
train.use_dora: false
train.use_galore: false
train.use_llama_pro: false
train.use_pissa: false
train.use_rslora: false
train.use_swanlab: false
train.val_size: 0
train.warmup_steps: 0
```


Issue:

The model's performance after fine-tuning(F1: 0.67) is significantly worse than the results reported in the paper(F1: 0.86 Qwen 2 VL 7B).

I suspect the problem might be related to the prompt construction, especially the format and the instruction phrasing. 


Request for Help:

- [ ] Could you please check if my prompt design for binary classification is suitable?

- [ ] Are there any recommended strategies to improve performance for binary classification tasks?


Thanks!

**Comment by JingbiaoMei** [comment]
Hi,

I double-checked my config file and found out that we use different settings for the LoRA hyperparameter. I am really sorry to confuse you yesterday. 

here is my config file for the baseline SFT experiments:
```
### model
model_name_or_path: QWen/QWen2-VL-7B-Instruct

### method
stage: sft
do_train: true
finetuning_type: lora
lora_target: all
lora_alpha: 256
lora_dropout: 0.05
lora_rank: 128
#deepspeed: examples/deepspeed/ds_z3_config.json
### dataset
dataset: harmp # video: mllm_video_demo
template: qwen2_vl
cutoff_len: 1024 
max_samples: 400000 # 400000
overwrite_cache: true
preprocessing_num_workers: 16

### output
output_dir: checkpoints/qwen2_vl-7b/lora/harmp/${DATE}_baseline
logging_steps: 10
save_steps: 5000
plot_loss: true
overwrite_output_dir: true

### train
per_device_train_batch_size: 16
gradient_accumulation_steps: 1
learning_rate: 1.0e-5
num_train_epochs: 6.0
lr_scheduler_type: cosine
warmup_ratio: 0.1
bf16: true
pure_bf16: true
ddp_timeout: 180000000
flash_attn: fa2
enable_liger_kernel: true

### eval
val_size: 0.001
per_device_eval_batch_size: 1
eval_strategy: steps
eval_steps: 5000


report_to: wandb
run_name: ${RUN_NAME}


```
Additionally,  to help you debug, I have uploaded the stage-1 LMM-RGCL trained checkpoint, you could load it with the base Qwen2-VL class. Since the stage-2 requires separate classifier code, you may first use this stage-1 checkpoint to help me debug the evaluation settings. 

https://huggingface.co/Jingbiao/Qwen2-VL-7B-Harm-P-LMM-RGCL

The prompt I use are fairly basic:
```
This is an image with: \"{}\" written on it. Is it hateful?
```
You should replace {} with the text caption. Our findings are:
- If you are fine-tuning the model, most prompts gives quite similar results.
- However, if zero-shot or few-shot is what you are interested, then some careful prompt engineering is crucial.

**Comment by NJIT-Fan** [comment]
Thank you for the clarification, and I really appreciate the response! You have done a lot of comparative experiments, very solid work!

**Comment by NJIT-Fan** [comment]
Hi,

I’m running into some issues and was hoping you could help me out.

I’ve aligned the prompt and parameter settings with the baseline, but it seems like the model starts overfitting after around 1000 steps. The training loss drops to about 0.02, but the average accuracy is only around 0.56 (on Qwen 2 VL 7B and Qwen 2.5VL 7B), which is much lower than expected.

I’ve checked the dataset, and it seems fine on my end, but I’m still not sure what’s going wrong.
Here’s the link to my dataset — would you mind taking a look when you have a chance?
👉 [Link](https://drive.google.com/drive/folders/1Z2jIqErTWvJ6aNa9nfJVXcWc1APNOMeC?usp=sharing)

Here is my code to transfer it to llama factory dataset format:
```python
import json

image_base_path = "/mmfs1/project/cs638/fy34/project_meme/dataset/harm/HarMeme/Harm-P/harmeme_images_us_pol/"

allowed_labels = {"not harmful", "somewhat harmful", "very harmful"}

label_mapping = {
    "not harmful": "benign",
    "somewhat harmful": "hateful",
    "very harmful": "hateful"
}

input_file_path = "./Harm-P/test_v1.jsonl"
data = []
with open(input_file_path, "r", encoding="utf-8") as f:
    for line in f:
        data.append(json.loads(line))

converted_data = []
for item in data:
    image_path = image_base_path + item["image"]

    text_content = item.get("text", "")

    human_value = f'<image> This is an image with: "{text_content.strip()}" written on it. Is it hateful?'
    
    filtered_labels = [label for label in item.get("labels", []) if label in allowed_labels]
    
    if filtered_labels:
        if any(label_mapping[label] == "hateful" for label in filtered_labels):
            final_label = "hateful"
        else:
            final_label = "benign"
    else:
        final_label = "benign"
        print("No allowed label found for an item; defaulting to 'benign'.")

    converted_item = {
        "conversations": [
            {
                "from": "human",
                "value": human_value
            },
            {
                "from": "gpt",
                "value": final_label
            }
        ],
        "images": [image_path]
    }
    converted_data.append(converted_item)

output_file_path = "./test.json"
with open(output_file_path, "w", encoding="utf-8") as f:
    json.dump(converted_data, f, ensure_ascii=False, indent=2)

print("Conversion completed, saved to:", output_file_path)
```

I’d really appreciate your help!

Best regards,

**Comment by JingbiaoMei** [comment]
[sharegpt_HarmP_train_instructblip.json](https://github.com/user-attachments/files/19410727/sharegpt_HarmP_train_instructblip.json)
Try this to help you debug.

**Comment by NJIT-Fan** [comment]
Thanks a lot! I'll try it right away.

**Comment by NJIT-Fan** [comment]
I have finished my experiments and ensured that my formatting aligned with yours. However, the average accuracy only reached **0.55**, while the paper reports an accuracy of **80+** for **SFT Qwen 2 VL 2B**. It might not be an issue with the data format. Below, I've attached some of the training logs from my process:

This is from your Hugging Face logs:

```json
{
  "classification_loss": 0.7851,
  "epoch": 0.015894039735099338,
  "grad_norm": 21.516714096069336,
  "learning_rate": 1.0619469026548673e-06,
  "lm_loss": 1.9864,
  "loss": 2.7715,
  "step": 3
}
```
And this is mine:
```json
{
epoch: 0.21739130434782608
grad_norm: 3.596904993057251
learning_rate: 0.000003603603603603604
loss: 0.5285
step: 40
}
```

It seems that I cannot find errors.

Regarding SFT, this should just be basic LoRA fine-tuning, without adding the loss function mentioned in the paper. Am I right?
 
I’d really appreciate your help!

<img width="1882" alt="Image" src="https://github.com/user-attachments/assets/5ee35485-8740-4034-9898-24c1c1c8d9a0" />

Best regards,

**Comment by JingbiaoMei** [comment]
> Regarding SFT, this should just be basic LoRA fine-tuning, without adding the loss function mentioned in the paper. Am I right?
Yes, SFT is the baseline method. 

![Image](https://github.com/user-attachments/assets/5ce774aa-a125-4821-9724-8a9038c5bf68)
This is my training log for the harmp lora sft model, it looks quite similar to me.

I think the issue would be the evaluation scripts. How do you evaluate the model, do you just extract yes/no from the answer?
Could you run the lora weights I shared with you and score that checkpoint?

**Comment by NJIT-Fan** [comment]
Thanks a lot! Yes, I'm using yes and no for the answers. By the way, how are you doing the evaluation? Are you loading the model and LoRA weights directly with the `transformers` library and then feeding the input for evaluation? I can try running an evaluation with the weights you shared to see if the results or generations look off.

**Comment by JingbiaoMei** [comment]
> By the way, how are you doing the evaluation? Are you loading the model and LoRA weights directly with the transformers library and then feeding the input for evaluation?

Yes, and you should expect same result no matter merging the weights or loading it directly with transformers library. 
The code should look like this:
```
            assert self.base_model_path is not None
            self.model = Qwen2VLForConditionalGeneration.from_pretrained(
                self.base_model_path,
                torch_dtype=torch.bfloat16,
                attn_implementation="flash_attention_2",
                device_map="auto",
                quantization_config=quantization_config
            ).eval()
            self.model.load_adapter(self.model_path)
            if self.max_pixels is None:
                self.processor = AutoProcessor.from_pretrained(self.base_model_path)
            else:
                self.processor = AutoProcessor.from_pretrained(self.base_model_path, max_pixels=self.max_pixels)
             
```

**Comment by NJIT-Fan** [comment]
My model still has the same results, after testing it with the transformers library. 
Anyway, Thank you, I plan to try it myself. I appreciate your advice!

**Comment by JingbiaoMei** [comment]
No worries! Hopefully, you'll get it working soon.

I plan to release the code very soon—likely within the next two days.

Let me know if you need any help in the meantime!

**Comment by NJIT-Fan** [comment]
Could you help me check this? I think I made a mistake somewhere in the pipeline. 

I used LLaMA Factory to load your LoRA weights and then do inference. I’ve carefully ensured that my dataset has the same format and content as yours. However, I’m only getting an F1 score and accuracy of 0.81.

Could you help me figure out what might be going wrong?

Here is my evaluation code:
```python
import json
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

def evaluate_classification(y_true, y_pred, average='weighted'):
    results = {}
    results['accuracy'] = accuracy_score(y_true, y_pred)
    results['precision'] = precision_score(y_true, y_pred, average=average, zero_division=0)
    results['recall'] = recall_score(y_true, y_pred, average=average, zero_division=0)
    results['f1_score'] = f1_score(y_true, y_pred, average=average, zero_division=0)
    results['confusion_matrix'] = confusion_matrix(y_true, y_pred)
    results['classification_report'] = classification_report(y_true, y_pred, zero_division=0)
    return results

def extract_data_from_jsonl(filepath):
    y_true = []
    y_pred = []
    with open(filepath, 'r', encoding='utf-8') as file:
        for line in file:
            line = line.strip()
            if not line:
                continue 
            record = json.loads(line)
            y_true.append(record['label'])
            y_pred.append(record['predict'])
    return y_true, y_pred

if __name__ == "__main__":
    filepath = "./saves/Qwen2-VL-7B/lora/author_eval/generated_predictions.jsonl"
    
    y_true, y_pred = extract_data_from_jsonl(filepath)
    metrics = evaluate_classification(y_true, y_pred)
    
    print("accuracy:", metrics['accuracy'])
    print("precision:", metrics['precision'])
    print("recall:", metrics['recall'])
    print("F1:", metrics['f1_score'])
    print("confusion_matrix:\n", metrics['confusion_matrix'])
    print("classification_report:\n", metrics['classification_report'])
```

<img width="737" alt="Image" src="https://github.com/user-attachments/assets/73fcabda-de79-4ff9-b3d2-28f5cb04e3cd" />

I’ve attached my dataset in JSON format and the code I’m using. I'd really appreciate it if you could take a quick look.

[test.json](https://github.com/user-attachments/files/19490681/test.json)

Thank you so much!

Best,

**Comment by JingbiaoMei** [comment]
Hi,

I've released the training and evaluation code for reproducing the experiments.
You may check if there is any mismatch in the pipelines.

**Comment by NJIT-Fan** [comment]
Thank you so much! I will try it!

**Comment by NJIT-Fan** [comment]
<img width="812" alt="Image" src="https://github.com/user-attachments/assets/6d8f07fa-9e79-4df5-9707-4c0dbef4b688" />

我好像还是没有训练出来，我是按照开源的代码做的。我只修改了路径。我也不确定我哪里出错了，可以加个好友问一下吗？
训练数据集已经完全对齐。
测试验证的数据集是开源的原始数据集。
能加个微信好友吗？
我的WeChat ID: iwillcomeback4u
非常感谢！

**Comment by NJIT-Fan** [comment]
<img width="604" alt="Image" src="https://github.com/user-attachments/assets/33cb83d1-ab9d-4b40-a41a-97434c9e3e21" />

Successfully reproduced. Thank you very much! I found that the only thing different from mine and yours is the image size. 224*224. I use the original size of the image. I appreciate it!

**Comment by JingbiaoMei** [comment]
> I found that the only thing different from mine and yours is the image size. 224*224. I use the original size of the image. 
We use the specific max pixels to reduce the number of tokens of the image, as in some datasets, the image is very high-definition and takes up thousands of tokens. However, I am still surprised that the image size makes such a difference.

**Comment by NJIT-Fan** [comment]
Yes. I didn't expect it to have such a big impact. Fixed image size seems to help the results a lot. Even qwen 2vl seems to have technology for dynamic image size.

**Comment by NJIT-Fan** [comment]
<img width="1458" alt="Image" src="https://github.com/user-attachments/assets/68b4b75b-5b22-498d-97b5-eac81243d14d" />

When I try to fine-tune on the PrideMM dataset. I meet this issue. Do you meet this issue before? If you meet this, How you fix it?
